### PR TITLE
[SDL 0256] Refactor Fuel Information Related Vehicle Data

### LINKED ISSUE #250: [SDL 0256] Refactor Fuel Information Related Vehicle Data
Proposal: [Refactor Fuel Information Related Vehicle Data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0256-Refactor-Fuel-Information-Related-Vehicle-Data.md)

> This proposal is to consolidate Fuel Information related vehicle data so that it supports multiple types of propulsion sources used by the vehicle.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/854

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with revisions.  The author will make the following changes to the proposal: add `KILOGRAM` to `CapacityUnit`; change maxValue of `level` and `capacity` in `FuelRange` to larger upper limit (1000000); fix `fuelLevel` and `fuelLevel_State` typos; update descriptions of `fuelLevel` and `fuelLevel_State` to state the param is deprecated as of x.x.x of the rpc spec, please see `fuelRange`; modify description of `fuelRange` param to be more descriptive of its extended capabilities.  The author can work with the Project Maintainer to finalize this description.

> The proposal .md file was updated to reflect these revisions on 11/14/19.

### PR DESCRIPTION
**Pull request by MSavitsk**
Fixes #250

This PR is **[ready]** for review.

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
Covered by existing unit tests.

### Summary
Applied [[SDL 0256] Refactor Fuel Information Related Vehicle Data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0256-Refactor-Fuel-Information-Related-Vehicle-Data.md) changes

##### Enhancements
* Added required getters/setters in related RPC classes

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) and this pull request adheres to the [Contributing Guide](./CONTRIBUTING.md).

### CODE REVIEW

**Comment by MSavitsk** [comment]
@kshala-ford @mrapitis @atiwari9 please review the changes.

**Comment by santhanamk** [comment]
@MSavitsk I reviewed your PR.  Below, are my suggestions.

**Source code changes:**

FuelRange.js
 - setType, setRange, setLevel shows up in SDL.min.js however the newly added setLevelState. setCapacity, and setCapacityUnit do not show up there.  Is that an issue or is it okay?

VehicleDataType.js
 - are you able to indicate that VEHICLEDATA_FUELLEVEL and VEHICLEDATA_FUELLEVEL_STATE are valid until X.x?

**Unit Test changes:**

_Test.js:_
 - Does Test.js need the enum `CapacityUnit`?

Can you add unit tests for CapacityUnit?

**Comment by vladmu** [comment]
@santhanamk thank you for review.

Regarding your points:

> VehicleDataType.js
> * are you able to indicate that VEHICLEDATA_FUELLEVEL and VEHICLEDATA_FUELLEVEL_STATE are valid until X.x?

The provided changes were generated by the RPC generator based on changes in the `rpc_spec/MOBILE_API.xml`, eg. https://github.com/smartdevicelink/rpc_spec/pull/262/files#diff-745a89443809bf31b6e5533025d37475R5851
So we expect that all required indications should be there before putting them into JS code.

>_Test.js:_
> * Does Test.js need the enum `CapacityUnit`?

I will check the existing tests and add required or answer here a little bit later.

**Comment by santhanamk** [comment]
Sounds good @vladmu.  Please let me know when you want me to review the code again.

**Comment by vladmu** [comment]
@santhanamk We've added unit tests, please review.

**Comment by vladmu** [comment]
> I reviewed the unit test files. Test.js and CapacityUnitTests.js look good.
> 
> For FuelRangeTests.js, can you also include the parameters type and range in your tests?

Hi @santhanamk, in this particular case it is not a problem to add those 2 parameters, but could you advise must I cover all class parameters in other cases? Sometimes they have a lot of existing parameters, where only a couple were changed by the proposal.

**Comment by santhanamk** [comment]
> > I reviewed the unit test files. Test.js and CapacityUnitTests.js look good.
> > For FuelRangeTests.js, can you also include the parameters type and range in your tests?
> 
> Hi @santhanamk, in this particular case it is not a problem to add those 2 parameters, but could you advise must I cover all class parameters in other cases? Sometimes they have a lot of existing parameters, where only a couple were changed by the proposal.

Hi @vladmu.  I think to be consistent, it is a good idea to cover all class parameters in other cases.

**Comment by vladmu** [comment]
@santhanamk @atiwari9 I'd like to inform you that if we need the full coverage of affected classes we need more time for unit tests and it would be impossible to provide all required changes today.

**Comment by santhanamk** [comment]
> @santhanamk @atiwari9 I'd like to inform you that if we need the full coverage of affected classes we need more time for unit tests and it would be impossible to provide all required changes today.

@vladmu I am ok if the unit tests just cover whatever was changed/added in the proposal.

**Comment by atiwari9** [comment]
@vladmu  @santhanamk  - We should cover tests related to changes for this proposal only. Adding other test cases is not in scope of current proposal. That is just good to have option.

**Comment by vladmu** [comment]
@santhanamk please write if I still require to add type and range in FuelRangeTests.js. If it is no more required, please review and approve this PR. Thank you!

**Comment by santhanamk** [comment]
> @santhanamk please write if I still require to add type and range in FuelRangeTests.js. If it is no more required, please review and approve this PR. Thank you!

You can add type and range in FuelRangeTests.js.

**Comment by vladmu** [comment]
> You can add type and range in FuelRangeTests.js.
Done

**Comment by santhanamk** [comment]
Hi @theresalech this PR is ready for Livio review.

**Comment by crokita** [comment]
While we continue our review of this PR, please pull in the new changes from the develop branch. We are removing the built files from the project so that merge conflicts will stop happening with them and the processing of contributing and reviewing code will be easier.
@MSavitsk

**Comment by vladmu** [comment]
@crokita the branch is refreshed with new changes from develop. Please review.

File: lib/js/src/rpc/messages/OnVehicleData.js [hunk 1]
```diff
@@ -123,7 +123,8 @@ class OnVehicleData extends RpcNotification {
 
     /**
      * Set the FuelLevel
-     * @param {Number} level - The fuel level in the tank (percentage) - The desired FuelLevel.
+     * @param {Number} level - The fuel level in the tank (percentage). This parameter is deprecated starting RPC Spec - The desired FuelLevel.
+     * 6.2, please see fuelRange.
```
[the remark this answers is not in the document]
Comment (vladmu) [reply lib/js/src/rpc/messages/OnVehicleData.js]: @crokita done in https://github.com/smartdevicelink/sdl_javascript_suite/pull/262/commits/c9bbafa8f078104966880656f530865b7c67b92e

**Pull request merged**